[hip] Docs: Overhaul HW implementation page (#1994)
ROCm/rocm-systems: projects/hip/docs/how-to/performance_guidelines.rst

* Avoid launching too many kernels (causes resource contention)

Multiprocessor level
--------------------

Maximize parallel execution within each multiprocessor:

* Ensure sufficient resident warps for every clock cycle
* Exploit instruction-level parallelism within warps
* Exploit thread-level parallelism across warps
* Balance resource usage for optimal occupancy

.. _memory optimization:

Memory throughput optimization
==============================

The first step in maximizing memory throughput is to minimize low-bandwidth
data transfers between the host and the device.

Additionally, maximize the use of on-chip memory (shared memory and caches) and 
minimize transfers with global memory.

.. _data transfer:

Data transfer optimization
--------------------------

**Minimize host-device transfers**

* Move computations from host to device when possible
* Create, use, and discard intermediate data structures on device
* Avoid unnecessary copies to host memory

**Batch small transfers**

Each memory transfer incurs fixed overhead from driver calls and PCIe transaction 
setup. Consolidating many small transfers into a single large transfer amortizes 

> this overhead across more data, achieving much higher effective bandwidth.

Reviewer: ```suggestion
Each memory transfer incurs a fixed overhead from driver calls and PCIe transaction setup. Consolidating many small transfers into a single large transfer amortizes this overhead across more data, resulting in much higher effective bandwidth.
```

Mapped page-locked memory allows zero-copy access, where the GPU reads directly 
from host memory without requiring an explicit transfer, eliminating redundant copies.

.. code-block:: cuda

   int integrated;
   hipDeviceGetAttribute(&integrated, hipDeviceAttributeIntegrated, device);
   if (integrated) {
       // Use mapped page-locked memory - no explicit copy needed
       hipHostMalloc(&ptr, size, hipHostMallocMapped);
   }

.. _device memory access:

Device memory access
--------------------

**Ensure proper alignment**

Memory hardware loads data in aligned chunks (typically 128 bytes). Using 
naturally aligned data types ensures each access maps to a single memory 
transaction, maximizing bandwidth and avoiding split transactions.

.. code-block:: cuda

   // Use naturally aligned types
   float4 data;  // 16-byte aligned
   float2 data;  // 8-byte aligned
   
   // Ensure structure alignment
   struct __align__(16) MyStruct {
       float4 data;
   };

**Optimize 2D array access**

Padding 2D arrays to multiples of the wavefront size ensures each row starts 
at an aligned memory boundary. This allows consecutive threads accessing the 

> same row to generate coalesced memory transactions, maximizing bandwidth.

Reviewer: ```suggestion
same row to generate coalesced memory transactions, thereby maximizing bandwidth.
```

.. code-block:: cuda

   // Ensure array width is multiple of warp size
   int width = ((actual_width + warpSize - 1) / warpSize) * warpSize;
   hipMalloc(&array, width * height * sizeof(float));
   
   // Access pattern
   int idx = x + width * y;  // width should be warp-aligned

**Coalesce memory accesses**

When consecutive threads in a wavefront access consecutive memory addresses, 
the hardware combines these into a single wide transaction. Non-coalesced 
patterns require multiple transactions, reducing effective bandwidth.

.. code-block:: cuda

   // Good: consecutive threads access consecutive addresses
   int idx = threadIdx.x + blockIdx.x * blockDim.x;
   data[idx] = value;
   
   // Bad: strided access
   int idx = threadIdx.x * stride;  // Non-coalesced if stride > 1
   data[idx] = value;

For understanding memory coalescing theory, see :ref:`memory_hierarchy_theory`.

**Use shared memory for data reuse**

Shared memory (LDS) provides low-latency on-chip storage shared across threads 
in a block. Loading data into shared memory once and reusing it many times 

> reduces global memory traffic, particularly effective for tiled algorithms like matrix multiplication.

Reviewer: ```suggestion
reduces global memory traffic, particularly effective for tiled algorithms such as matrix multiplication.
```

.. code-block:: cuda

   __global__ void optimized_kernel(float* input, float* output) {
       __shared__ float tile[TILE_SIZE][TILE_SIZE];
       
       // Load data into shared memory
       tile[threadIdx.y][threadIdx.x] = input[...];
       __syncthreads();
       
       // Reuse data from fast shared memory
       float result = 0;
       for (int i = 0; i < TILE_SIZE; i++) {
           result += tile[threadIdx.y][i] * tile[i][threadIdx.x];
       }
       __syncthreads();
       
       output[...] = result;
   }

**Avoid bank conflicts in shared memory**

Shared memory is organized into banks that can service one request per cycle. 
When multiple threads in a warp access the same bank simultaneously, they 
serialize, reducing throughput. Padding arrays by one element shifts addresses 

> to avoid systematic conflicts.

Reviewer: ```suggestion
Shared memory is organized into banks, each capable of servicing one request per cycle. When multiple threads in a warp access the same bank simultaneously, the requests are serialized, reducing throughput. Padding arrays by one element shifts addresses to avoid systematic conflicts.
```

bitwise operations (shifts, AND, OR) are single-cycle instructions on integer 
units, making them far more efficient than equivalent arithmetic for power-of-two calculations.

.. code-block:: cuda

   // Prefer multiplication over division
   float result = value * 0.5f;     // Fast
   float result = value / 2.0f;     // Slower
   
   // Use bitwise operations for powers of 2
   int index = threadIdx.x << 2;    // Multiply by 4
   int mask = (1 << n) - 1;         // Create bit mask

**Use single-precision when possible**

AMD GPUs have significantly higher throughput for single-precision (FP32) 
operations compared to double-precision (FP64). Using single-precision math 
functions can deliver substantial performance gains when FP64 accuracy is not required.

.. code-block:: cuda

   // Single-precision (faster)
   float result = sinf(x);
   float result = expf(x);
   
   // Double-precision (slower, use only when necessary)
   double result = sin(x);
   double result = exp(x);

**Leverage fast math intrinsics**

Hardware-specific intrinsics bypass some accuracy checks and use lookup tables 
or polynomial approximations, trading slight precision loss for significantly 

> higher throughput. Use these when the application can tolerate reduced precision.

Reviewer: ```suggestion
Hardware‑specific intrinsics bypass certain accuracy checks and use lookup tables or polynomial approximations, trading slight precision loss for significantly higher throughput. These should be used when the application can tolerate reduced precision.
```
